src/trajectory.jl

```diff
@@ -565,19 +575,37 @@ function isterminated(h::Hamiltonian, t::BinaryTree{<:ClassicNoUTurn})
 end
 
 """
-    isterminated(h::Hamiltonian, t::BinaryTree{<:GeneralisedNoUTurn})
+    isterminated(
+        h::Hamiltonian, 
+        t::BinaryTree{<:GeneralisedNoUTurn},
+        tleft::BinaryTree{<:GeneralisedNoUTurn}, 
+        tright::BinaryTree{<:GeneralisedNoUTurn}
+    )
 
 Detect U turn for two phase points (`zleft` and `zright`) under given Hamiltonian `h`
 using the generalised no-U-turn criterion.
 
-Ref: https://arxiv.org/abs/1701.02434
+Ref: https://arxiv.org/abs/1701.02434 https://github.com/stan-dev/stan/pull/2800
 """
-function isterminated(h::Hamiltonian, t::BinaryTree{<:GeneralisedNoUTurn})
-    # z0 is starting point and z1 is ending point
-    z0, z1 = t.zleft, t.zright
+function isterminated(
+    h::Hamiltonian, 
+    t::BinaryTree{<:GeneralisedNoUTurn}, 
+    tleft::BinaryTree{<:GeneralisedNoUTurn}, 
+    tright::BinaryTree{<:GeneralisedNoUTurn}
+)
     rho = t.c.rho
-    s = (dot(rho, ∂H∂r(h, -z0.r)) >= 0) || (dot(-rho, ∂H∂r(h, z1.r)) >= 0)
-    return Termination(s, false)
+    s1 = generalised_uturn_criterion(rho, ∂H∂r(h, t.zleft.r), ∂H∂r(h, t.zright.r))
+
+    rho = tleft.c.rho + tright.zleft.r
+    s2 = generalised_uturn_criterion(rho, ∂H∂r(h, t.zleft.r), ∂H∂r(h, tright.zleft.r))
+
+    rho = tleft.zright.r + tright.c.rho
+    s3 = generalised_uturn_criterion(rho, ∂H∂r(h, tleft.zright.r), ∂H∂r(h, t.zright.r))
+    return Termination(s1 || s2 || s3, false)
+end
+
+function generalised_uturn_criterion(rho, p_sharp_minus, p_sharp_plus)
```

Comment:
It's good now that we abstract this check out so that doing what I proposed in https://github.com/TuringLang/AdvancedHMC.jl/pull/207#issuecomment-655436469 would be easy as well.